curiosrc/market/deal_ingest.go

```diff
@@ -19,81 +22,268 @@ import (
 	"github.com/filecoin-project/lotus/chain/types"
 	"github.com/filecoin-project/lotus/curiosrc/seal"
 	"github.com/filecoin-project/lotus/lib/harmony/harmonydb"
+	lpiece "github.com/filecoin-project/lotus/storage/pipeline/piece"
 )
 
+var log = logging.Logger("piece-ingestor")
+
+const loopFrequency = 10 * time.Second
+
 type Ingester interface {
 	AllocatePieceToSector(ctx context.Context, maddr address.Address, piece api.PieceDealInfo, rawSize int64, source url.URL, header http.Header) (api.SectorOffset, error)
 }
 
 type PieceIngesterApi interface {
+	ChainHead(context.Context) (*types.TipSet, error)
 	StateMinerInfo(context.Context, address.Address, types.TipSetKey) (api.MinerInfo, error)
 	StateMinerAllocated(ctx context.Context, a address.Address, key types.TipSetKey) (*bitfield.BitField, error)
 	StateNetworkVersion(ctx context.Context, key types.TipSetKey) (network.Version, error)
 }
 
-type PieceIngester struct {
-	db  *harmonydb.DB
-	api PieceIngesterApi
+type openSector struct {
+	number        abi.SectorNumber
+	currentSize   abi.PaddedPieceSize
+	earliestEpoch abi.ChainEpoch
+	index         uint64
+	lk            sync.Mutex
+	openedAt      time.Time
+	sealNow       bool
 }
 
-func NewPieceIngester(db *harmonydb.DB, api PieceIngesterApi) *PieceIngester {
-	return &PieceIngester{db: db, api: api}
+type PieceIngester struct {
+	ctx                 context.Context
+	db                  *harmonydb.DB
+	api                 PieceIngesterApi
+	miner               address.Address
+	mid                 uint64 // miner ID
+	windowPoStProofType abi.RegisteredPoStProof
+	sectorSize          abi.SectorSize
+	sealRightNow        bool // Should be true only for CurioAPI AllocatePieceToSector method
+	maxWaitTime         time.Duration
+	forceSeal           chan struct{}
+	openSectors         []*openSector
```

Comment:
I'd really like to think of ways to put this and coordinate pieces through the database; This will be relevant to us also when we move in markets into Curio, and ideally everything about piece ingest would be HA as well.

Reply:
Ack, I will move everything to DB. It does allow for better co-ordination and HA.